blockdag/reachability.go

```diff
@@ -0,0 +1,501 @@
+}
+
+// This method counts the size of each subtree under this node.
+// The method outcome is exactly equal to the following recursive
+// implementation:
+//
+// func (rtn *reachabilityTreeNode) countSubtreesUp() uint64 {
+//     subtreeSize := uint64(0)
+//     for _, child := range rtn.children {
+//         subtreeSize += child.countSubtreesUp()
+//     }
+//     return subtreeSize
+// }
+//
+// However, we are expecting (linearly) deep trees, and so a
+// recursive stack-based approach is inefficient and will hit
+// recursion limits. Instead, the same logic was implemented
+// using a (queue-based) BFS method. At a high level, the
+// algorithm uses BFS for reaching all leaves and pushes
+// intermediate updates from leaves via parent chains until all
+// size information is gathered at the root of the operation
+// (i.e. at rtn).
+//
+// Note the role of the subtreeSize field in the algorithm.
+// For each node rtn this field is initialized to 0. The field
+// has two possible states:
+// * rtn.subtreeSize > |rtn.children|:
+//   this indicates that rtn's subtree size is already known and
+//   calculated.
+// * rtn.subtreeSize <= |rtn.children|:
+//   we are still in the counting stage of tracking which of
+//   rtn's children has already calculated its subtree size.
+//   This way, once rtn.subtree_size = |rtn.children| we know we
+//   can pull subtree sizes from children and continue pushing
+//   the readiness signal further up
+func (rtn *reachabilityTreeNode) countSubtreesUp() uint64 {
+	queue := []*reachabilityTreeNode{rtn}
+	for len(queue) > 0 {
+		var current *reachabilityTreeNode
+		current, queue = queue[0], queue[1:]
+		if len(current.children) == 0 {
+			// We reached a leaf
+			current.subtreeSize = 1
+		}
+		if current.subtreeSize <= uint64(len(current.children)) {
+			// We haven't yet calculated the subtree size of
+			// the current node. Add all its children to the
+			// queue
+			for _, child := range current.children {
+				queue = append(queue, child)
+			}
+			continue
+		}
+
+		// We reached a leaf or a pre-calculated subtree.
+		// Push information up
+		for current != rtn {
+			current = current.parent
+			current.subtreeSize++
+			if current.subtreeSize != uint64(len(current.children)) {
+				// Not all subtrees of the current node are ready
+				break
+			}
+			// All subtrees of current have reported readiness.
+			// Count actual subtree size and continue pushing up.
+			childSubtreeSizeSum := uint64(0)
+			for _, child := range current.children {
+				childSubtreeSizeSum += child.subtreeSize
+			}
+			current.subtreeSize = childSubtreeSizeSum + 1
+		}
+	}
+	return rtn.subtreeSize
+}
+
+// applyIntervalDown applies new intervals using a BFS traversal.
+// The intervals are allocated according to subtree sizes and the
+// 'split' allocation rule (see the split() method for further
+// details)
+func (rtn *reachabilityTreeNode) applyIntervalDown(interval *reachabilityInterval) error {
```

Comment:
Up and down are very confusing, please use "past" and "future" instead.

Update everywhere, not only here.